files/zh-cn/web/api/cookiechangeevent/index.md

```diff
@@ -0,0 +1,64 @@
+---
+title: CookieChangeEvent
+slug: Web/API/CookieChangeEvent
+l10n:
+  sourceCommit: 74206b3f81736bac558470f36222544cc67ba9e2
+---
+
+{{securecontext_header}}{{APIRef("Cookie Store API")}}
+
+{{domxref("Cookie Store API", "", "", "nocode")}} 的 **`CookieChangeEvent`** 接口是发生任何 cookie 更改时在 {{domxref("CookieStore")}} 上触发的 {{domxref("CookieStore/change_event", "change")}} 事件的事件类型。
+
+导致派发 `CookieChangeEvent` 的 Cookie 变更包括：
+
+- 新创建一个 cookie 且没有立刻被删除。此用例中 `type` 为 "changed"。
+- 新创建一个 cookie 且立刻被删除。此用例中 `type` 为 "deleted"。
+- cookie 被删除。此用例中 `type` 为 "deleted"。
+
+> **备注：** cookie 被新插入的具有相同名称、域和路径的另一个 cookie 替换时，并且不会触发更改事件。
```

Comment:
https://wicg.github.io/cookie-store/#process-changes

> A cookie change is a cookie and a type (either changed or deleted):
> 
> - A cookie which is removed due to an insertion of another cookie with the same name, domain, and path is ignored.
> - A newly-created cookie which is not immediately evicted is considered changed.
> - A newly-created cookie which is immediately evicted is considered deleted.
> - A cookie which is otherwise evicted or removed is considered deleted

第一个问题：这里与 w3c 的原文表述存在不一致，原文说的是这种行为是 “删除” 或 “变更”。并没说该事件的类型属性为 "deleted"。 
第二个问题：备注的内容属于判断条件之一。

---
此外，以下 TEST CASE 并未成功验证 w3c 的逻辑。

前置条件：

```yml
ua: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36'
```

```js
cookieStore.addEventListener("change", (event) => {
  console.log(event);
});
```

用例 1：

```js
cookieStore.set("cookie1","cookie1-value");
cookieStore.delete("cookie1");
```

用例 2：

```js
cookieStore.set({
  name: "cookie1",
  value: "cookie1-value",
  domain: "developer.mozilla.org",
  path: '/',
});
cookieStore.set({
  name: "cookie1",
  value: "cookie1-value2",
  domain: "developer.mozilla.org",
    path: '/',
});
```

用例 1 仍触发了两次事件
用例 2 没有 ignore 变更

Reply:
`type` 貌似始终为 `"change"`。可以给 mdn/content 提 issue 或 PR。这里的描述确实和实现不相符。

Reply:
w3c 中该段描述的是 CookieChange 的 type。并不是 CookieChangeEvent 的 type。

----
window 环境下伪代码：

```ts
type CookieChange = {
  type: 'changed' | 'deleted';
  cookie: Cookie;
}

const cookieChanges = new Set<CookieChange>();

const fireCookieChangeEvent = () => {
  if (!cookieChanges.size) continue;

  const changed = [];
  const deleted = [];
  for (let change of cookieChanges) {
     if (change.type === 'changed') {
       changed.push(change.cookie)
     } else if (change.type === 'deleted') {
       deleted.push(change.cookie)
     }
  }
  
  cookieChanges.clear();
  
  CookieStore.emit('change', new CookieChangeEvent('change', {
    changed,
    deleted,
  }));
  
  requestIdleCallback(fireCookieChangeEvent);
}

requestIdleCallback(fireCookieChangeEvent);
```

Reply:
> > 用例 1 仍触发了两次事件
> 
> 这个条件下可能不是“立即删除的情况”，因为 CookieStore 的方法都是异步的，调用两个方法可能不符合“立即”的概念
> 
> 考虑观察如使用同步的 `document.cookie` 的情况

强调“立即”的原因正是因为 “异步” 和异步后可以批处理任务。
与 React （Fiber 架构后版本）多次 setState 只触发一次 render 类似。

Reply:
> 用例 1 仍触发了两次事件

有可能与 cookieStore api remove 实现有关系。
cookieStore.remove 的实现是修改其过期时间而不是从 cookie list 移除元素。
只能看 blink 的实现验证了

Reply:
```js
cookieStore.set({
    name: 'cookie321',
    value: '22',
    expires: Date.now() - 1000
})
// CookieChangeEvent.deleted contain the cookie.
```

设置一个“已过期” 的 cookie 能达到 “立即 evicted”

Reply:
看起来这种情况才是真正的“立即删除”（添加时设置 cookie 过期时间已过）

> ```js
> cookieStore.set({
>     name: 'cookie321',
>     value: '22',
>     expires: Date.now() - 1000
> })
> // CookieChangeEvent.deleted contain the cookie.
> ```

不管是 document.cookie 还是 CookieStore，先添加后删除都不属于“立即删除”

Reply:
> 不管是 document.cookie 还是 CookieStore，先添加后删除都不属于“立即删除”

我倒认为该实验特性未完全按 w3c 标准实现。
至少目前为止，我试出 batch update 的场景。（CookieChangeEvent.changed 或 CookieChangeEvent.deleted > 1）